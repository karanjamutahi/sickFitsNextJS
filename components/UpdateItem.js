import React, { Component } from 'react';
import { Mutation, Query } from 'react-apollo';
import gql from 'graphql-tag';
import Router from 'next/router';

import Form from './styles/Form';
import formatMoney from '../lib/formatMoney';
import ErrorMessage from './ErrorMessage';

const SINGLE_ITEM_QUERY = gql`
    query SINGLE_ITEM_QUERY(
        $id: ID!
    ) {
        item(where: {
            id: $id
        }) {
            title
            price
            description
            image
            largeImage
        }
    }
`;

const UPDATE_ITEM_MUTATION = gql`
    mutation UPDATE_ITEM_MUTATION(
        $id: ID!
        $title: String
        $description: String
        $price: Int
        $image: String
        $largeImage: String
    ) {
        updateItem(
        id:$id
        title: $title
        description: $description
        price: $price
        image: $image
        largeImage: $largeImage
        ) {
            id
            title
            description
            price
            image
            largeImage
        }
    }
`;

class UpdateItem extends Component {
    state = {};

    handleChange = (e) => {
        const { name,type,value } = e.target;
        const val = type === 'number' ? parseFloat(value) : value;
        this.setState({ [name]: val});
    };

    updateItem = async (e, updateItemMutation) => {
        e.preventDefault();
        const res = await updateItemMutation({
            variables: {
                id: this.props.id,
                ...this.state,

            }
        });
    };

    render(){
        return(
            <Query query={SINGLE_ITEM_QUERY} variables={{ id: this.props.id }}>
                {({ data } ,loading) => {
                    if(loading) return <p>Loading ...</p>;
                    if(!data.item) return <h1>No Item Found</h1>;
                    return (
                        <Mutation mutation={UPDATE_ITEM_MUTATION} variables={this.state}>
                            {(updateItem, { loading, error})=>(
                                
                            <Form onSubmit= {e => this.updateItem(e, updateItem)}>
                            <h2>Update {data.item.title} Details</h2>
                            <ErrorMessage error={error}></ErrorMessage>
                            <fieldset disabled={loading} aria-busy={loading}>

                                {data.item.image && <img src={data.item.image} alt="Image Preview"></img> /**TODO: Add an 'x' button to delete entry from cloudinary and upload a new image */}
                                {/*
                                <label htmlFor="file">
                                    Upload Image
                                    <input type="file" name="file" id="file" placeholder='Upload Image' required onChange={this.handleImage}/>
                                </label>
                                */}
                                <label htmlFor="title">
                                    Title
                                    <input type="text" name="title" id="title" placeholder='Title' required defaultValue={data.item.title} onChange={this.handleChange}/>
                                </label>

                                <label htmlFor="price">
                                    Price (KSH)
                                    <input type="number" name="price" id="price" placeholder='Price' required defaultValue={data.item.price} onChange={this.handleChange}/>
                                </label>

                                <label htmlFor="description">
                                    Description
                                    <textarea type="number" name="description" id="description" placeholder='Enter A Description' required defaultValue={data.item.description} onChange={this.handleChange}/>
                                </label>

                                <button type='Submit'>Sav{loading ? 'ing' : 'e'} Changes</button>
                            </fieldset>
                        </Form>
                        )}
                    </Mutation>
                )
            }}
        </Query>
        );
    }
};

export default UpdateItem;
export { UPDATE_ITEM_MUTATION };
export { SINGLE_ITEM_QUERY };