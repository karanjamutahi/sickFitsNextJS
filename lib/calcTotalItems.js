function totalItems(cart) {
    return cart.reduce((tally, cartItem) => tally + cartItem.quantity, 0)
}

export default totalItems;