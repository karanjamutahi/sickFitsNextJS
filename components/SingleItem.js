import React, { Component } from 'react';
import styled from 'styled-components';
import Head from 'next/head';
import {  Query } from 'react-apollo';

import { SINGLE_ITEM_QUERY } from './UpdateItem';
import Error from './ErrorMessage';
import Loading from './Loading';

const SingleItemStyles = styled.div`
  max-width: 1200px;
  margin: 2rem auto;
  box-shadow: ${props => props.theme.bs};
  display: grid;
  grid-auto-columns: 1fr;
  grid-auto-flow: column;
  min-height: 800px;
  img {
    width: 100%;
    height: 100%;
    object-fit: contain;
  }
  .details {
    margin: 3rem;
    font-size: 2rem;
  }
`;

class SingleItem extends Component {
    render() {
        return (
            <Query query={SINGLE_ITEM_QUERY} variables={{id: this.props.id}}>
            {({ data , error, loading })=>{
                    if (loading) return <Loading loading={true}></Loading> ;
                    if(error) return <Error error={error}></Error>;
                    if(!data.item) return <p>404: Item not Found</p> ;
                    return (
                        <SingleItemStyles>
                            <Head>
                                <title>Sick Fits | {data.item.title}</title>
                            </Head>
                            <img src={data.item.largeImage} alt={data.item.title} />
                            <div className="details">
                                <h2>Viewing {data.item.title}</h2>
                                <p>{data.item.description}</p>
                            </div>
                        </SingleItemStyles>
                    )
                }
            }
        </Query>
        );
    }
}

export default SingleItem;