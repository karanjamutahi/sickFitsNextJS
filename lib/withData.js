import withApollo from 'next-with-apollo';
import ApolloClient from 'apollo-boost';
import { devEndpoint, prodEndpoint } from '../config';
import { LOCAL_STATE_QUERY } from '../components/Cart';

function createClient({ headers }) {
  return new ApolloClient({
    uri: process.env.NODE_ENV === 'development' ? devEndpoint : devEndpoint,
    request: operation => {
      operation.setContext({
        fetchOptions: {
          credentials: 'include',
        },
        headers,
      });
    },
    clientState: {
      resolvers: {
        Mutation: {
          toggleCart(_, variables, { cache }) {
            //read cartopen value from cache
             const { cartOpen } = cache.readQuery({
               query: LOCAL_STATE_QUERY,
             });
             //Write the cart State to the opposite
             const data = {
               data: {
                 cartOpen: ! cartOpen
               }
             };
             cache.writeData(data);
             return data; 
          }
        }
      },
      defaults: {
        cartOpen: false,
      }
    }
  });
}

export default withApollo(createClient);
