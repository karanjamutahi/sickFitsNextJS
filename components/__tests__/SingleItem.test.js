import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import SingleItem from '../SingleItem';
import { SINGLE_ITEM_QUERY } from '../UpdateItem';
import { MockedProvider } from 'react-apollo/test-utils';
import { fakeItem } from '../../lib/testUtils';

describe("Single Item Component", () => {
    let mocks = [
        {
            request: { query: SINGLE_ITEM_QUERY, variables: { id: fakeItem().id }},
            result: { data: { item: fakeItem(), },}
        }
];

    it("Renders with proper data", async () => {
        const wrapper = mount(
        <MockedProvider mocks={mocks}>
            <SingleItem id={fakeItem().id}></SingleItem>
        </MockedProvider>
        );
        expect (wrapper.find('Loading')).toBeTruthy();
        await wait();
        wrapper.update();
        expect(toJSON(wrapper.find('h2'))).toMatchSnapshot() ;
    });

    it("Errors with a not found item", async () => {
        mocks[0].result = {
            errors: [
                {
                    message: "Item not found!"
                }
            ]
        };

        const wrapper = mount(
            <MockedProvider mocks={mocks}>
                <SingleItem id={fakeItem().id}></SingleItem>
            </MockedProvider>
        );

        await wait();
        wrapper.update();
        const item = toJSON(wrapper.find('[data-test="graphql-error"]'));
        expect(item).toMatchSnapshot();
    });
});