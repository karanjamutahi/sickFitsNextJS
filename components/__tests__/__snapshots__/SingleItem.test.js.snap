// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Single Item Component Errors with a not found item 1`] = `
<p
  data-test="graphql-error"
>
  <strong>
    Shoot!
  </strong>
  Item not found!
</p>
`;

exports[`Single Item Component Renders with proper data 1`] = `
<h2>
  Viewing 
  dogs are best
</h2>
`;
